docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,569 @@
+
+Developer-Centric Approach
+
+* Prioritize clear, intuitive interfaces and best-practice design principles.
+* Improve Developer Experience: Provide tools, samples, and best practices to foster an efficient and user-friendly development environment.
+
+Leverage Extensibility
+
+* Utilize the module's modular design to support a wide range of cryptographic tools, key types, and methods, ensuring adaptability for future technological advancements.
+* Integrate support for advanced cryptographic features, ensuring the module's position at the forefront of cryptographic technologies.
+
+Quality Assurance
+
+* Enhanced Test Coverage: Improve testing methodologies to ensure the robustness and reliability of the module.
+* Conduct an Audit: After implementation, perform a comprehensive audit to identify potential vulnerabilities and ensure the module's security and stability.
+
+## Technical Goals
+
+As technical goals, the aim is to create a robust, flexible, and future-proof cryptographic module. This is achieved through the following key points:
+
+Wide Hardware Device & Cloud-based HSM Interface Support:
+
+* Design a foundational interface for various hardware devices (Ledger, YubiKey, Thales, etc.) and cloud-based HSMs (Amazon, Azure) to cater to both current and future implementations.
+
+Plugin Architecture and Dependency Injection
+
+* Establish the architectural foundation for an extensible plugin system and integrate a dependency injection framework, ensuring modularity, testability, and third-party integrations.
+
+* Design an environment for plugin testing, ensuring developers can validate integrations without compromising system integrity.
+
+Interface considerations
+
+* Design should take into considerations support for Trusted Platform Module (TPM) 2.0 and similar devices to anticipate future enhancements.
+
+* Design should take into account the Cryptographic Token Interface Standard (PKCS#11)
+
+Increase cryptographic versatility
+
+* Support for a broad spectrum of cryptographic techniques
+* Extend support for more hash functions (e.g. pedersen, argon2, Argon2d/I/id, Blake3, etc.)
+* Extend support for more signature schemes (e.g. secp256r1, ed25519, ed448, sr25519, etc.)
+* More advanced methods ( Post-Quantum Cryptography (PQC) methods
+* Threshold signatures and encryption
+
+Community Engagement Infrastructure:
+
+* Structure the design with tools and documentation interfaces in mind, enabling a seamless future rollout of resources for developer engagement.
+
+## Proposed architecture
+
+### Introduction
+
+In the proposed architecture, each package is decoupled and isolated. Adding new implementations consist of implementing the required interfaces.
+
+```mermaid
+classDiagram
+
+Keyring <|-- Wallet
+
+SecureStorage <|-- Keyring
+SecureItem <|-- SecureStorage
+CryptoProvider <|-- SecureItem
+
+Hasher <|-- CryptoProvider
+Verifier <|-- CryptoProvider
```

Comment:
how come client side code needs to have a hasher, verifier and cipher? i might be missing the usecase

Reply:
The idea is that the `CryptoProvider` has the entire crypto toolchain available. Maybe on the specific case of the sdk's `/client` might not be directly used but for example if another zone wants to build something that actually needs a `Verifier` on their client, they have it out of the box without needing to import external packages.